### DISCUSSION TITLE
Replay Packets with their parameter values

### DISCUSSION THREAD

**Discussion by EmbeddedChristian** [yamcs/yamcs#767] [Q&A] [answered]
Hello,

I am currently setting up the YAMCS-Quickstart project for communication with a piece of software that will run on a remote system. I have setup the database to work with a non-CCSDS packet, since that was required by our client (they gave us a custom format). I have a simulator running on my system that periodically sends me packets and I can receive those and read out the Parameter-Values no problem. I can even send Telecommands to the simulator no problem.

Our client now asked for a way to replay the received packets in order of reception and to display their respective Parameter Values. When I start a replay in YAMCS I can see the received packets in order of reception, but I can only see their binary data. I can also look at every single parameter in the Telemetry->Parameter section. They are replayed correctly but I don't see a way of displaying the packet they were received from. 

I found the packet-viewer in the YAMCS documentation, but I could not find any information on how it works nor how to set it up. It seems to be part of the YAMCS main repository, so I would make a guess that it is part of the YAMCS backend. Is there a way to set it up to work with the YAMCS quickstart project? And does it even do, what I want it to do?

TLDR: Is there a way to display packets with their respective parameter-values in a replay?

Kind regards,
Christian

**Comment by xpromache** [comment, marked as answer]
The Packet Viewer is a java swing application. 

You can run it in the quickstart example if you add this in the dependency section of pom.xml
```xml
               <dependency>
                        <groupId>org.yamcs</groupId>
                        <artifactId>packet-viewer</artifactId>
                        <version>${yamcsVersion}</version>
                </dependency>
```

And then you can start it like this:
```sh
mvn yamcs:run-tool -D=yamcs.tool=org.yamcs.ui.packetviewer.PacketViewer
```
I'm not sure how difficult is to connect it to a replay though. It connects to a stream and when you do a replay there is a replay stream but you would need to find its name. Please give it a try, if you manage to start it and it works in realtime I can maybe help you make it work in replay.

**Reply by EmbeddedChristian** [reply]
Hello,

Thanks for the fast reply. I will give it a try at monday and get back to you asap.

**Reply by ckoch-rst-rostock** [reply]
Hello, I am writing from my account at work now (I am @EmbeddedChristian).

When adding the dependency and running the command I can't seem to start the packet-viewer in my main-project, getting the following error:
```
No columns definition found in ui preferences
Exception in thread "main" java.lang.NullPointerException
	at org.yamcs.ui.packetviewer.PacketViewer.getRecentFiles(PacketViewer.java:1076)
	at org.yamcs.ui.packetviewer.PacketViewer.updateMenuWithRecentFiles(PacketViewer.java:495)
	at org.yamcs.ui.packetviewer.PacketViewer.installMenubar(PacketViewer.java:406)
	at org.yamcs.ui.packetviewer.PacketViewer.<init>(PacketViewer.java:272)
	at org.yamcs.ui.packetviewer.PacketViewer.main(PacketViewer.java:1239)
```

When starting the packet-viewer in a fresh quick-start project it works flawlessly. I can then connect to my YAMCS instance and see all the TM packets coming in. 

Is there a way to give the packet-viewer instance access to my MDB for offline access of Dump-files if I start it like this? When I try to open files I seem to miss the mdb.yaml and the database, but I can't seem to figure out where they should be located.

Edit: In your answer you stated "(...) It connects to a stream and when you do a replay there is a replay stream but you would need to find its name (...)". All I could find so far was the name of the replay-processor. 

Also: Thank you for initial answer, it already helped me a lot.

**Reply by xpromache** [reply]
And to answer to the question of offline access - the packet-viewer can load an MDB definition from ~/.yamcs/mdb.yaml and it can process files offline based on that.
You have to setup ~/.yamcs/packet-viewer.yaml defining the packetInputStream (how to split your dump-files into packets) and your packet pre-processor. 

Those files you can also put in the etc folder and Fabian just reminds me that we release also a packet-viewer tgz in the standard release https://github.com/yamcs/yamcs/releases

The tgz has an etc directory inside but in order to use it with your custom pre-processor (and possibly other java algorithms if you reference any in you MDB), you will have to copy your jar in the lib directory. I think if you want to give that package away to someone to use it standalone, it's better you go that way (then they will not require maven).

**Reply by ckoch-rst-rostock** [reply]
The tgz works well for me, thanks for the idea, but I have an issue when trying to use my custom pre-processor. I keep getting a ClasNotFound Exception:
```
Mar 20 14:06:29.695 org.yamcs.ui.packetviewer.PacketViewer [1] [SEVERE] Cannot instantiate the packet preprocessor: Cannot instantiate object from class com.example.myproject.MyPacketPreprocessor: java.lang.ClassNotFoundException: com.example.myproject.MyPacketPreprocessor
	org.yamcs.utils.YObjectLoader.loadObject(YObjectLoader.java:94)
	org.yamcs.ui.packetviewer.PacketViewer.loadPacketPreprocessor(PacketViewer.java:1307)
	org.yamcs.ui.packetviewer.PacketViewer.readConfig(PacketViewer.java:1318)
	org.yamcs.ui.packetviewer.PacketViewer.<init>(PacketViewer.java:179)
	org.yamcs.ui.packetviewer.PacketViewer.main(PacketViewer.java:1264)
Caused by: java.lang.ClassNotFoundException: com.example.myproject.MyPacketPreprocessor
	java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
	java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	java.base/java.lang.Class.forName0(Native Method)
	java.base/java.lang.Class.forName(Class.java:315)
	org.yamcs.utils.YObjectLoader.loadObject(YObjectLoader.java:29)
	org.yamcs.ui.packetviewer.PacketViewer.loadPacketPreprocessor(PacketViewer.java:1307)
	org.yamcs.ui.packetviewer.PacketViewer.readConfig(PacketViewer.java:1318)
	org.yamcs.ui.packetviewer.PacketViewer.<init>(PacketViewer.java:179)
	org.yamcs.ui.packetviewer.PacketViewer.main(PacketViewer.java:1264)
```
Which is weird, because I inserted my jar into lib/ext and that is part of the ClassPath, which I double checked. Do I have to include other files as well, or do I need to setup the Jar in a specific manner? I also tried to insert it into lib/ but I got the same result.

**Reply by xpromache** [reply]
You can  run the shell script using bash -x to check if your jar gets added to the classpath.

**Reply by ckoch-rst-rostock** [reply]
It seems like my jar-files were broken. I re-created them and then it worked fine, thank you for your help

**Comment by xpromache** [comment]
You must be running an old version, this issue was fixed here:
https://github.com/yamcs/yamcs/commit/da1056c522f4d19ca879ca8087c4c96a230ebd00

**Reply by ckoch-rst-rostock** [reply]
Ok, I was running 5.6.0. I changed to 5.7.10 now.

The packet-viewer now starts correctly, but when I try to connect to YAMCS I get an error:
```
org.yamcs.xtce.BooleanDataType; local class incompatible: stream classdesc SerialVersionUID = 1, local class SerialVerisionUID = 2
```

**Reply by xpromache** [reply]
Is your server running an old Yamcs then?

**Reply by ckoch-rst-rostock** [reply]
My bad, forgot to restart it. It works now.
